### PR TITLE
Fixes #3232 Classifications do not cover whole file

### LINKED ISSUE #3232: Classifications do not cover whole file
Sometimes while editing you will lose all syntax highlighting from a certain point to the end of the file. We should not do this.

This is likely due to changes in #3224

### PR DESCRIPTION
**Pull request by zooba**
Fixes #3232 Classifications do not cover whole file
Moves in-proc tokenization into PythonTextBufferInfo
Switches some users of the classifier to the correct token source
Fixes line number management of token cache (i.e. don't trust the tokens)
Removes index comparisons from SourceLocation
Fixes explicit/incorrect uses of three argument SourceLocation constructor

### CODE REVIEW

**Comment by zooba** [comment]
I bit the bullet and centralized tokenization/caching, which I've been worrying about having to do for a while. It means we can stop using the classifier and use the PythonTextBufferInfo in a variety of places (which is better because we control the lifetime of the text buffer info, so it'll clean up certain race conditions with VS). But there's different information available, so it's a bit more work to translate them all.

This is probably also the last piece we need to get off ReverseExpressionParser. But I'm starting to think it may need a bit of a rewrite from what's here in order to keep snapshots in sync. Feel free to review, but there may be more changes before it can be merged.

**Comment by zooba** [comment]
This is probably ready for review now. I'm running the test suite against it to try and flush out issues before merging though. But it's essentially all internal refactoring with no visible impact (though the use of the new event for tokenization should improve typing responsiveness a bit).

File: Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs [hunk 1]
```diff
@@ -455,6 +579,94 @@ sealed class PythonTextBufferInfo {
...
+            for (int lineNo = firstLine; lineNo < span.Snapshot.LineCount; ++lineNo) {
+                var line = span.Snapshot.GetLineFromLineNumber(lineNo);
+
+                var beforeState = lineTokenization.State;
+
+                tokenCache[lineNo] = lineTokenization = TokenizeLine(tokenizer, line, beforeState);
+
+                var afterState = lineTokenization.State;
+
+                // stop if we visted all affected lines and the current line has no tokenization state
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs]: np typo 'visted'

File: Python/Product/PythonTools/PythonTools/TokenCache.cs [hunk 2]
```diff
@@ -54,13 +53,136 @@ internal struct LineTokenization : ITag {
...
+        /// 0-based index where the token starts on the line.
+        /// </summary>
+        public int Column;
+        /// <summary>
+        /// Number of characters included in the token.
+        /// </summary>
+        public int Length;
+    }
+
+    public struct TrackingTokenInfo {
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/TokenCache.cs]: hard to tell from here, does this have to be public?

Comment (zooba) [reply Python/Product/PythonTools/PythonTools/TokenCache.cs]: Strictly no, but then we have a stack of internal API on PythonTextBufferInfo. We can do that, but I think there's value in that API being public. But maybe it's better just to hide the whole thing for now...

Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/TokenCache.cs]: PythonTextBufferInfo isn't public either. I wouldn't worry about making this public if it's not necessary.

Comment (zooba) [reply Python/Product/PythonTools/PythonTools/TokenCache.cs]: Ah, you're right. Guess I misread the "sealed" as "public" at some point.

**Pull request merged**